### PR TITLE
added nanvar to numpy frontend

### LINKED ISSUE #7454: nanvar

### PR DESCRIPTION
**Pull request by SaifAshraf70**
Closes #7454

### CODE REVIEW

**Comment by SaifAshraf70** [comment]
Hi @kurshakuz could you review my pull request

File: ivy/functional/frontends/numpy/statistics/averages_and_variances.py [hunk 1]
```diff
@@ -158,7 +150,7 @@ def cov(
             # if w exists, use weighted average
             xw = x.multiply(w)
             w_sum = ivy.sum(w)
-            average = ivy.stable_divide(ivy.sum(xw, axis=1) , w_sum)
+            average = ivy.stable_divide(ivy.sum(xw, axis=1), w_sum)
```
Comment (kurshakuz) [inline ivy/functional/frontends/numpy/statistics/averages_and_variances.py:153]: There is no need to change another function in your PR. If you want to fix other things than adding nanvar, please create separate PR

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hey! Thank you for your contribution! Your test produces AssertionErrors, can you please fix them? Thanks

File: ivy/functional/frontends/numpy/statistics/averages_and_variances.py (hunk 1 above)
Comment (SaifAshraf70) [reply ivy/functional/frontends/numpy/statistics/averages_and_variances.py:153]: > There is no need to change another function in your PR. If you want to fix other things than adding nanvar, please create separate PR

This happened automatically due to the pre-commit framework

**Review by kurshakuz — COMMENTED** [review]
Looks good to me now! Can you please show the outpt of the test_numpy_nanvar? Once it passes I will merge it. Thanks!

**Comment by SaifAshraf70** [comment]
> Looks good to me now! Can you please show the outpt of the test_numpy_nanvar? Once it passes I will merge it. Thanks!

Hi @kurshakuz I tried a lot in the assertion errors but all my tries failed and I think the problem is from the ivy core functions itself not my code because I ran my code before and all tests were passed.
Could you help me in this error, I will send the error message to you on Discord.

**Comment by kurshakuz** [comment]
> Hi @kurshakuz I tried a lot in the assertion errors but all my tries failed and I think the problem is from the ivy core functions itself not my code because I ran my code before and all tests were passed. Could you help me in this error, I will send the error message to you on Discord.

Please formulate your problem concretely, I cannot help you that way. Check line by line in a debugger mode and let me know which functions exactly produce errors. If you are still stuck, ask your issue in the corresponding Discord channel as (1) you will get response much faster (2) can get a response from a person who had similar issue before (3) will help in the future to someone who gets stuck with that problem too 🙂

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hi! I have checked your implementation locally and it still fails for that test:
```
ret_np = array([nan], dtype=float16), ret_from_gt_np = array([inf], dtype=float16), rtol = 0.1, atol = 0.1
ground_truth_backend = 'TensorFlow'
```

```
========================================================= short test summary info ==========================================================
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.numpy-False-False]
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.jax-False-False]
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.tensorflow-False-False]
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.torch-False-False]
===================================================== 4 failed, 171 warnings in 16.19s =====================================================
```

**Comment by SaifAshraf70** [comment]
> Hi! I have checked your implementation locally and it still fails for that test:
> 
> ```
> ret_np = array([nan], dtype=float16), ret_from_gt_np = array([inf], dtype=float16), rtol = 0.1, atol = 0.1
> ground_truth_backend = 'TensorFlow'
> ```
> 
> ```
> ========================================================= short test summary info ==========================================================
> FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.numpy-False-False]
> FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.jax-False-False]
> FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.tensorflow-False-False]
> FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.torch-False-False]
> ===================================================== 4 failed, 171 warnings in 16.19s =====================================================
> ```

Hello @kurshakuz every time I run my code on my local machine the 4 tests passed and this is the output:
============================= test session starts ==============================
collecting ... collected 4 items

test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.numpy-False-False] 
test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.jax-False-False] 
test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.tensorflow-False-False] 
test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.torch-False-False] 

======================== 4 passed, 2 warnings in 46.58s ========================

Process finished with exit code 0
PASSED [ 25%]WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
PASSED [ 50%]PASSED [ 75%]2022-12-29 12:21:47.315177: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2022-12-29 12:21:47.315216: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
PASSED [100%]

**Comment by SaifAshraf70** [comment]
Please verify that your local repo is up to date with the original repo

**Comment by SaifAshraf70** [comment]
Could you tell me what should I do now?
All the tests passed on my local machine
@kurshakuz

**Comment by SaifAshraf70** [comment]
Hello, @kurshakuz I solved the error and all the tests passed on my local machine. Thanks!

File: ivy/functional/frontends/numpy/statistics/averages_and_variances.py [hunk 2]
```diff
@@ -195,3 +195,41 @@ def cov(x, y=None, bias=False, dtype=None, fweights=None, aweights=None, ddof=No
...
+
+        ret = ivy.var(a, axis=axis, correction=ddof, keepdims=keepdims, out=out)
+
+        if ivy.is_array(where):
+            where = ivy.array(where, dtype=ivy.bool)
+            ret = ivy.where(where, ret, ivy.default(out, ivy.zeros_like(ret)), out=out)
+
+    all_nan = ivy.isnan(ret)
+    if ivy.all(all_nan):
+        ret = ivy.astype(ret, ivy.array([float("inf")]))
```
Comment (kurshakuz) [inline ivy/functional/frontends/numpy/statistics/averages_and_variances.py:234]: This line is currently producing that error:
```
E           ivy.exceptions.IvyBackendException: tensorflow: astype: Cannot convert the argument `type_value`: <tf.Tensor: shape=(1,), dtype=float64, numpy=array([inf])> to a TensorFlow DType.
```

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hey! Thank you for fixing it. I have tested your implementation once again and it now produces the following errors. Can you resolve them as well?

```
E     TypeError: Cannot convert the argument `type_value`: <tf.Tensor: shape=(1,), dtype=float64, numpy=array([inf])> to a TensorFlow DType.
```
```
E       AssertionError: the return with a TensorFlow backend produced data type of float16, while the return with a torch backend returned a data type of float64.
```

**Comment by SaifAshraf70** [comment]
> Hey! Thank you for fixing it. I have tested your implementation once again and it now produces the following errors. Can you resolve them as well?
> 
> ```
> E     TypeError: Cannot convert the argument `type_value`: <tf.Tensor: shape=(1,), dtype=float64, numpy=array([inf])> to a TensorFlow DType.
> ```
> 
> ```
> E       AssertionError: the return with a TensorFlow backend produced data type of float16, while the return with a torch backend returned a data type of float64.
> ```

Could you give me any hint to solve this error?
I think I need to use type promotion?

**Comment by kurshakuz** [comment]
I would suggest to first fix the TypeError. As you see from the message the issue is probably from the fact that you try to feed the unsupported data type to the TensorFlow DType. After fixing that issue we can look into the type promotion which can solve the Assertion error 🙂

**Comment by SaifAshraf70** [comment]
Hi @kurshakuz I solved the TypeError and the 4 tests passed I wish it pass on your local machine too. Thanks!

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hey! I have just pulled from the master branch and tested your implementation and it still fails locally.

Here is what I got:
```
args = (), kwargs = {'a': array([-1., -1.]), 'as_variable_flags': [False, False], 'atol': 0.1, 'axis': 0, ...}, values = (None, 0.0), ret = None
frontend_ret = 0.0, frontend_ret_np_flat = [array(0.)], ret_np_flat = []
```

Throws following error:
```
E       IndexError: list index out of range
E       Falsifying example: test_numpy_nanvar(
E           dtype_x_axis=(['float64'], [array([-1., -1.])], 0, 0),
E           dtype=['float64'],
E           where=[array([False, False])],
E           keep_dims=False,
E           as_variable=[False],
E           num_positional_args=0,
E           with_out=False,
E           native_array=[False],
E           fn_tree='ivy.functional.frontends.numpy.nanvar',
E           frontend='numpy',
E           on_device='cpu',
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AXicY2BkAAMoBaaR2cgAAAC/AAY=') as a decorator on your test case
```

I am not sure how tests pass for you.

Can you run 
```
pytest ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar
```
and paste the full console log here as file? I will then check it as well. Thanks! 🙂

**Comment by SaifAshraf70** [comment]
4c8f496dcdfc:python3 -u /opt/.pycharm_helpers/pycharm/_jb_pytest_runner.py --target test_averages_and_variances.py::test_numpy_nanvar
Testing started at 5:00 PM ...
Launching pytest with arguments test_averages_and_variances.py::test_numpy_nanvar --no-header --no-summary -q in /opt/project/ivy/ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics

Use Database in ReadOnly Mode with local caching !
============================= test session starts ==============================
collecting ... collected 4 items

test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.numpy-False-False] 
test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.jax-False-False] 
test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.tensorflow-False-False] 
test_averages_and_variances.py::test_numpy_nanvar[cpu-ivy.functional.backends.torch-False-False] 

======================== 4 passed, 2 warnings in 51.85s ========================

Process finished with exit code 0
PASSED [ 25%]WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
WARNING:root:NumPy does not support autograd, declaring a 'variable' is identical to declaring an 'array' when using numpy backend.
PASSED [ 50%]PASSED [ 75%]2023-01-05 15:01:04.203084: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2023-01-05 15:01:04.203161: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
PASSED [100%]

**Comment by SaifAshraf70** [comment]
> Hey! I have just pulled from the master branch and tested your implementation and it still fails locally.
> 
> Here is what I got:
> 
> ```
> args = (), kwargs = {'a': array([-1., -1.]), 'as_variable_flags': [False, False], 'atol': 0.1, 'axis': 0, ...}, values = (None, 0.0), ret = None
> frontend_ret = 0.0, frontend_ret_np_flat = [array(0.)], ret_np_flat = []
> ```
> 
> Throws following error:
> 
> ```
> E       IndexError: list index out of range
> E       Falsifying example: test_numpy_nanvar(
> E           dtype_x_axis=(['float64'], [array([-1., -1.])], 0, 0),
> E           dtype=['float64'],
> E           where=[array([False, False])],
> E           keep_dims=False,
> E           as_variable=[False],
> E           num_positional_args=0,
> E           with_out=False,
> E           native_array=[False],
> E           fn_tree='ivy.functional.frontends.numpy.nanvar',
> E           frontend='numpy',
> E           on_device='cpu',
> E       )
> E       
> E       You can reproduce this example by temporarily adding @reproduce_failure('6.55.0', b'AXicY2BkAAMoBaaR2cgAAAC/AAY=') as a decorator on your test case
> ```
> 
> I am not sure how tests pass for you.
> 
> Can you run
> 
> ```
> pytest ivy_tests/test_ivy/test_frontends/test_numpy/test_statistics/test_averages_and_variances.py::test_numpy_nanvar
> ```
> 
> and paste the full console log here as file? I will then check it as well. Thanks! 🙂

I think something went wrong on your machine because I didn't find this error at all

**Review by kurshakuz — APPROVED** [review]
You were right, I had some issues with my setup! Now all tests pass 🎉 
Thank you for your contribution, I will go ahead and merge your PR 🙂

**Comment by SaifAshraf70** [comment]
Thank you

> You were right, I had some issues with my setup! Now all tests pass 🎉 Thank you for your contribution, I will go ahead and merge your PR 🙂

I want to thank you very much for your help @kurshakuz ❤❤

**Pull request merged**